smb/smb.go

```diff
@@ -754,6 +816,554 @@ type QueryDirectoryRes struct {
 	Buffer             []byte
 }
 
+type QueryInfoReq struct {
+	Header
+	StructureSize         uint16 // Must always be 41 regardless of Buffer size
+	InfoType              byte
+	FileInfoClass         byte
+	OutputBufferLength    uint32
+	InputBufferOffset     uint16 `smb:"offset:Buffer"`
+	Reserved              uint16
+	InputBufferLength     uint32 `smb:"len:Buffer"`
+	AdditionalInformation uint32
+	Flags                 uint32
+	FileId                []byte `smb:"fixed:16"`
+	Buffer                []byte
+}
+
+type QueryInfoRes struct {
+	Header
+	StructureSize      uint16 // Must always be 9
+	OutputBufferOffset uint16 `smb:"offset:Buffer"`
+	OutputBufferLength uint32 `smb:"len:Buffer"`
+	Buffer             []byte
+}
+
+type SecurityDescriptor struct {
+	Revision    uint16
+	Control     uint16
+	OffsetOwner uint32
+	OffsetGroup uint32
+	OffsetSacl  uint32 // From beginning of struct?
+	OffsetDacl  uint32 // From beginning of struct?
+	OwnerSid    *SID
+	GroupSid    *SID
+	Sacl        *PACL
+	Dacl        *PACL
+}
+
+type PACL struct {
+	AclRevision uint16
+	AclSize     uint16
+	AceCount    uint32
+	ACLS        []ACE
+}
+
+// MS-DTYP Section 2.4.4.1 ACE_HEADER
+type ACEHeader struct {
+	Type  byte
+	Flags byte
+	Size  uint16 //Includes header size?
+}
+
+type ACE struct {
+	Header ACEHeader
+	Mask   uint32
+	Sid    SID //Must be multiple of 4
+}
+
+type SID struct {
+	Revision       byte
+	NumAuth        byte
+	Authority      []byte
+	SubAuthorities []uint32
+}
+
+const (
+	securityDescriptorOwner = "owner"
+	securityDescriptorGroup = "group"
+	securityDescriptorDACL  = "dacl"
+	securityDescriptorSACL  = "sacl"
+	securityDescriptorEnd   = "end"
+)
+
+func (self *SecurityDescriptor) MarshalBinary() (ret []byte, err error) {
```

Comment:
Implement BinaryMarshallable interface
`func (self *SecurityDescriptor) MarshalBinary() (ret []byte, err error) {`
should be
`func (self *SecurityDescriptor) MarshalBinary(meta *encoder.Metadata) (ret []byte, err error) {`